File: packages/backend/src/conversation/summary.ts
import type { ConversationTitles } from './title'
import type { AgentServer } from '@demicodes/agent'
import type { RunnerRegistry } from '../runner/registry'
import type { ControlService, ConversationRecord } from '../storage/control'
import type { ConversationStores } from '../storage/conversation-store'
import { resolveExecutionTarget } from './execution-target'

export interface ConversationSummaryDeps {
  stores: ConversationStores
  server: AgentServer
  control: ControlService
  registry: Pick<RunnerRegistry, 'deviceIdentity'>
  titles: Pick<ConversationTitles, 'generating'>
}

/**
 * A conversation as the browser lists it: its record, its live status, and
 * `cwd`, the directory its work runs in, resolved the same way for a
 * device directory, a workspace and the Cloud, so the browser never derives
 * it. Live activity wins; a cold unfinished checkpoint is interrupted,
 * never reported as still running.
 */
export async function conversationSummary(
  conversation: ConversationRecord,
  deps: ConversationSummaryDeps
) {
  const { stores, server, control, registry } = deps
  const cwd = (await resolveExecutionTarget(control, registry, conversation)).path
  const summary = stores.summary(conversation.id)
  const live = server.sessionPhase(conversation.id)
  let status:
    'running' | 'compacting' | 'interrupted' | 'error' |
    'stopped' | 'completed' | 'idle' = 'idle'
  if (server.treeActive(conversation.id))
    status = live === 'compacting' ? 'compacting' : 'running'
  else if (!live && summary.phase !== 'idle')
    status = 'interrupted'
  else if (summary.last === 'error')
    status = 'error'
  else if (summary.last === 'abort')
    status = 'stopped'
  else if (summary.last === 'response')
    status = 'completed'
  return {
    ...conversation,
    cwd,
    status,
    revision: summary.revision,
    unread: summary.revision > conversation.readRevision,
    // Whether asking for a new title could say anything the last one did not (`product.md` § Conversation titles).
    titleCurrent: conversation.userMessages <= conversation.titledMessages,
    titleGenerating: deps.titles.generating(conversation.id),
  }
}

File: crates/backend/src/conversation/summary.rs
//! Conversations as the browser lists them (`web-api.md` § Sidebar
//! mutations, read state and page synchronization): each record with the
//! directory its work runs in, its status from the live tree or else from its
//! last checkpoint, and its output revision. The persisted facts come from a
//! read-only connection, so listing hundreds of conversations takes no
//! writer from a running one.

use demi_core::SessionPhase;
use demi_web_api::conversations::{ConversationStatus, ConversationSummary};
use futures_util::future::try_join_all;

use super::root_of;
use crate::shard::Shard;
use crate::storage::StorageError;
use crate::storage::conversation_index::ConversationRecord;
use crate::storage::tree::{self, SummaryFacts, Terminal};

impl Shard {
    /// The summaries of the user's conversations that are archived, or that
    /// are not, in sidebar order.
    pub(crate) async fn conversation_summaries(&self, archived: bool) -> Result<Vec<ConversationSummary>, StorageError> {
        let records = self.services().control.conversations(self.user().clone(), archived).await?;
        try_join_all(records.into_iter().map(|record| self.conversation_summary(record))).await
    }

    /// `record` as the browser lists it.
    pub(crate) async fn conversation_summary(&self, record: ConversationRecord) -> Result<ConversationSummary, StorageError> {
        let facts = self
            .services()
            .conversations
            .read(&record.id, tree::summary)
            .await?
            .unwrap_or(SummaryFacts::EMPTY);
        let live = self
            .agent()
            .tree(&root_of(&record.id))
            .map(|tree| (tree.is_quiescent(), tree.root().session().phase()));
        let status = status(live, &facts);
        let cwd = self.resolve_target(&record).await?.path().to_owned();
        let (provider_id, model_id) = match record.model {
            Some(model) => (Some(model.provider), Some(model.model)),
            None => (None, None),
        };
        Ok(ConversationSummary {
            unread: facts.revision > record.read_revision,
            id: record.id,
            title: record.title,
            archived: record.archived,
            pinned: record.pinned,
            read_revision: record.read_revision,
            target: record.target,
            context_version: record.context_version,
            provider_id,
            model_id,
            created_at: record.created_at,
            updated_at: record.updated_at,
            cwd,
            status,
            revision: facts.revision,
        })
    }
}

/// A conversation's status: running or compacting while its live tree will
/// go on working without the user, interrupted when its checkpoint was saved
/// in a turn and no tree is live, else what its latest terminal block says.
fn status(live: Option<(bool, SessionPhase)>, facts: &SummaryFacts) -> ConversationStatus {
    match live {
        Some((false, SessionPhase::Compacting)) => return ConversationStatus::Compacting,
        Some((false, _)) => return ConversationStatus::Running,
        None if facts.phase != SessionPhase::Idle => return ConversationStatus::Interrupted,
        Some((true, _)) | None => {}
    }
    match facts.last {
        Some(Terminal::Response) => ConversationStatus::Completed,
        Some(Terminal::Error) => ConversationStatus::Error,
        Some(Terminal::Abort) => ConversationStatus::Stopped,
        None => ConversationStatus::Idle,
    }
}
